docs/trace/extending-the-sdk/MyFilteringProcessor.cs

```diff
@@ -18,15 +18,14 @@
 using System.Diagnostics;
 using OpenTelemetry;
 
-internal class MyFilteringProcessor : BaseProcessor<Activity>
+internal class MyFilteringProcessor : CompositeProcessor<Activity>
```

Comment:
Purely from the reader's perspective - I guess folks would question "why would I need to derive from CompositeProcessor?".

Reply:
🤔 this is a good question, and I can't think of simple explanation to give users. The alternative to making `SetParentProvider` public is problematic for similar reasons. I'd struggle to come up with a concise way to explain to users under what circumstances they need to be sure to call `SetParentProvider`.

In either case the explanation is basically: if you're newing up a processor and you're not directly passing it to `providerBuilder.AddProcessor()` (i.e., it's nested in another processor or something) then you gotta make sure that the parent provider gets set - which in turn ensures the provider has access to the configured resource... 

but this will raise questions of why the parent provider needs to be set and what does it need access to a resource

Reply:
I don't have a good solution for this! Agree with @alanwest exposing `SetParentProvider` feels wrong. I just added some XML comments explaining (or hinting at) what is going on in the example. That might help?

We could introduce a class to make it more self-explanatory like...

```csharp
public class DelegatingProcessor<T> : CompositeProcessor<T>
{
   public DelegatingProcessor(BaseProcessor<T> innerProcessor)
      : base(new[] { innerProcessor ?? throw new ArgumentNullException(nameof(innerProcessor) })
   {
   }
}
```

...or...

```csharp
    public class DelegatingProcessor<T> : BaseProcessor<T>
    {
        public DelegatingProcessor(BaseProcessor<T> innerProcessor)
        {
            Guard.ThrowIfNull(innerProcessor);

            this.InnerProcessor = innerProcessor;
        }

        protected BaseProcessor<T> InnerProcessor { get; }

        public override void OnStart(T data)
        {
            this.InnerProcessor.OnStart(data);
        }

        public override void OnEnd(T data)
        {
            this.InnerProcessor.OnEnd(data);
        }

        internal override void SetParentProvider(BaseProvider parentProvider)
        {
            this.InnerProcessor.SetParentProvider(parentProvider);
        }

        protected override bool OnForceFlush(int timeoutMilliseconds)
        {
            return this.InnerProcessor.ForceFlush(timeoutMilliseconds);
        }

        protected override bool OnShutdown(int timeoutMilliseconds)
        {
            return this.InnerProcessor.Shutdown(timeoutMilliseconds);
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                this.InnerProcessor.Dispose();
            }

            base.Dispose(disposing);
        }
    }
```

I doubt anyone will like that either, but it's an option 🤣

Reply:
> I doubt anyone will like that either, but it's an option 🤣

Yea not yet excited about this idea. I do agree the naming offers a slight improvement, but I think it's worth sitting on a bit. Definitely think it's worth continuing this discussion in hopes of finding a better solution.

Reply:
Something about this problem kinda feels similar to middleware branching. Like the filter processor effectively introduces a branch in the pipeline. Something like

```csharp
if (filterFunc(span)) {
    // invoke the next process in the chain
} else {
    // terminate the pipeline - maybe call a noop processor or something
}
```